### DISCUSSION TITLE
Power off Waydroid using Andoid's GUI, without shuting down the host.

### DISCUSSION THREAD

**Discussion by TriMoon** [waydroid/waydroid#1084] [Q&A] [answered]
See title, is this possible without using the "power off" button on the host?
I can't seem to find any settings to assign or change the buttons/keys used by Waydroid for this functionality.

Yes i know it automatically puts itself into sleep or that we can stop the session via command-line, but that's not what i'm after.

Best would be to have a menu option we could choose inside Waydroid's Android, just like in KDE's application menu.

**Comment by AndrewPrzelucki** [comment]
You need to install an app to achieve this. Search for "power" on F-Droid or "power button" on the Google Play Store. You will need to enable the app as a device administrator in settings->security. Make sure to use an app that does not require root privileges .

**Reply by TriMoon** [reply]
Thanks i will try to find what you suggest when i switch to Waydroid again. :+1: 
_(Kinda bussy with other more urgent stuff under X11 at moment.)_

# Update
I just tried a few but they are useless in combination with Waydroid, because the session keeps running while the container is stopped....

**Reply by unique-EJ** [reply]
> You need to install an app to achieve this. Search for "power" on F-Droid or "power button" on the Google Play Store. You will need to enable the app as a device administrator in settings->security. Make sure to use an app that does not require root privileges .

To quote _automation - How can I open the power menu in Android without pressing the power button?_ on Android Enthusiasts Stack Exchange,

> You can install Android Accessibility Suite from the Play Store.

I also recommend to use the Switch Access settings actions it has under accessibility to control the _Global actions_ with keyboard shortcuts. I assigned _Notifications_ `KEYCODE_MENU`, _Quick Settings_ `Shift+KEYCODE_MENU` and _Overview_ `Alt+KEYCODE_MENU` and it supports keyboard switches for selection, scanning and "touch and hold".

**Reply by roke-julian-lockhart** [reply]
> How can I open the power menu in Android without pressing the power button?

[`android.stackexchange.com/q/246297`][1] is the URI to it. [^1]

[^1]: [`android.stackexchange.com/review/suggested-edits/268238`][2]

[2]: https://android.stackexchange.com/review/suggested-edits/268238

[1]: https://android.stackexchange.com/q/246297/311449

**Reply by unique-EJ** [reply]
> > How can I open the power menu in Android without pressing the power button?
> 
> [`android.stackexchange.com/q/246297`](https://android.stackexchange.com/q/246297/311449) is the URI to it.

OK, you're welcome for the context reference.

**Comment by TriMoon** [comment, marked as answer]
My solution is to create a manual desktop file:

- `~/.local/share/applications/Waydroid.PowerOff.desktop`
```
[Desktop Entry]
Categories=X-WayDroid-App;
Comment=
Exec=waydroid session stop
GenericName=Shutdown Waydroid session
Icon=system-shutdown
Name=Waydroid PowerOff
NoDisplay=false
Path=
StartupNotify=true
Terminal=false
TerminalOptions=
Type=Application
X-KDE-SubstituteUID=false
X-KDE-Username=
X-Purism-FormFactor=Workstation;Mobile;
```
- This is a modified version of the default `Waydroid.desktop` file.

**Reply by oooo-ps** [reply]
pgrep waydroid still shows active process after `waydroid session stop`

**Reply by unique-EJ** [reply]
> pgrep waydroid still shows active process after `waydroid session stop`

What about `waydroid session stop && kill $(pgrep waydroid)`?
@oooo-ps , @TriMoon - example:

```
touch ~/.local/share/applications/Waydroid.PowerOff.desktop;

cat ~/.local/share/applications/Waydroid.PowerOff.desktop - <<END-OF-HERE-TEXT
[Desktop Entry]
Categories=X-WayDroid-App;
Comment=
Exec=/bin/sh -c 'waydroid session stop && systemctl stop waydroid-container'
GenericName=Shutdown Waydroid session
Icon=system-shutdown
Name=Waydroid Power-Off
NoDisplay=false
Path=
StartupNotify=true
Terminal=false
TerminalOptions=
Type=Application
X-KDE-SubstituteUID=false
X-KDE-Username=
X-Purism-FormFactor=Workstation;Mobile;
END-OF-HERE-TEXT
```
**but the kill signal needs root privileges and the Waydroid process should be stopped through _systemd_.**
Without `sudo `, `systemctl` should use a system authentication window and that will show on the Wayland desktop environment.

**Reply by unique-EJ** [reply]
To know it stops:
```
if { systemctl is-active waydroid-container | grep -x -q 'active'; }; then
  { waydroid session stop && { sudo systemctl stop waydroid-container; pgrep waydroid; }; };
fi
```

**Comment by roke-julian-lockhart** [comment]
> https://github.com/waydroid/waydroid/discussions/1084#discussion-5591008

@TriMoon, [`issues/1052`](https://github.com/waydroid/waydroid/issues/1052#issue-1850286079) tracks this.

**Comment by skidnik** [comment]
An alternative solution would be to edit the original `~/.local/share/applications/Waydroid.desktop` and add an action that can be triggered from the Waydroid icon's right click menu.

```[Desktop Entry]
Type=Application
Name=Waydroid
Exec=waydroid show-full-ui
Categories=X-WayDroid-App;
X-Purism-FormFactor=Workstation;Mobile;
Icon=waydroid
NoDisplay=false
#my additions start below
Actions=stop-session;

[Desktop Action stop-session]
Name=Stop session
Exec=waydroid session stop
```

**But** waydroid overwrites that entry each time it is started. So currently I have to duplicate the desktop entry, which may or may not be possible depending on the DE.
Would be nice if this action was added to `/usr/lib/waydroid/tools/services/user_manager.py`.

**Reply by roke-julian-lockhart** [reply]
> https://github.com/waydroid/waydroid/discussions/1084#discussioncomment-10635800

@skidnik, I think you should submit that as a PR.

**Reply by skidnik** [reply]
Now that you mentioned PRs I found that there's already #1356 .
